<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>BCB420 - Computational Systems Biology</title>
    <meta charset="utf-8" />
    <meta name="author" content="Ruth Isserlin" />
    <meta name="date" content="2020-01-06" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# BCB420 - Computational Systems Biology
## Course Introduction
### Ruth Isserlin
### 2020-01-06

---



class: center, middle

![](https://bcb.csb.utoronto.ca/files/2019/05/cropped-Header_artwork_new_WP_site_May2019-1.jpg)

Welcome to BCB420 Spring 2020 - &lt;br&gt;
Computational System Biology
&lt;br&gt;&lt;br&gt;
Instructor : Ruth Isserlin
&lt;br&gt;&lt;br&gt;
TA : Owen Whitley

???

---

# Course Overview

* Brief Introduction&lt;br&gt;
* Course aims&lt;br&gt;
* Marking breakdown&lt;br&gt;
* Where to find all information pertaining to the course&lt;br&gt;
* Quiz - January 21,2020 (Mock Quiz - January 14)&lt;br&gt;
* Wiki and Journal&lt;br&gt;
* Assignment #1 - Due January 28, 2020&lt;br&gt;
* Docker&lt;br&gt;

---

class: center
# Introduction

I am a Bioinformatician in [Gary Bader's lab]() &lt;br&gt;&lt;br&gt;
--
I have been at &lt;img src="https://cdn.freebiesupply.com/logos/large/2x/university-of-toronto-logo-png-transparent.png" width="5%" /&gt; for over 20 years&lt;br&gt;&lt;br&gt;
--
I have been working in the field of bioinformatics for almost the same amount of time. &lt;br&gt;&lt;br&gt;
--
I have worked with varying types of data including protein-protein interaction data, genomics, proteomics, metabolomics, methylation .... ,&lt;br&gt;&lt;br&gt;

--
I have worked with many different programming languages including c, Perl, java, python,matlab, R, good old bash...&lt;br&gt;&lt;br&gt;

--
What I have learned over all my years in bioinformatics is:&lt;br&gt;&lt;br&gt; 

--
The data **will** change, &lt;br&gt;
--
the methods **will** evolve and require increasing amounts of resources,&lt;br&gt;
--
but it will always come down to mapping identifiers :)
---

# Course Aims

What is Computational Systems Biology?&lt;br&gt;&lt;br&gt;
--
According to the U of T course description: &lt;br&gt;&lt;br&gt;
.center["Current approaches to using the computer for analyzing and modeling biology as integrated molecular systems."]&lt;br&gt;&lt;br&gt;

--

According to [Nature](https://www.nature.com/)'s 2002 article [Computational Systems Biology](Kitano, H. Computational systems biology. Nature 420, 206–210 (2002) doi:10.1038/nature01254) is:&lt;br&gt;&lt;br&gt;
.center["To understand complex biological systems requires the integration of experimental and computational research — in other words a systems biology approach. Computational biology, through pragmatic modelling and theoretical exploration, provides a powerful foundation from which to address critical scientific questions head-on."]&lt;br&gt;

---
class: left

# Course Aims

So, what are **we** going to be covering during this course?&lt;br&gt;
![](./images/course_overview_flowchart.png)

* Everyone is going to find and claim an unique -omics expression dataset.
* We will then normalize and clean the dataset
* We will then perform differential expression analysis on the dataset.
* We will the conduct pathway and network analysis on the dataset in order to visualize and interpret it. 

---
class: left

#Marking Breakdown

| Item | Weight | Due Date |
| --- | --- | --- |
| initial set up | 5 marks | January 10, 2020 |
| Quiz on preparatory material | 20 marks | January 21, 2020 |
| Assignment 1 | 20 marks | February 4, 2020 |
| Assignment 2 | 20 marks | March 3, 2020 |
| Assignment 3 | 20 marks | March 24, 2020 |
| Journal/Insights | 15 marks | On going. |
| Total | 100 marks | - |


---
class: center
background-image: url(./images/course_setup.png)
background-size: contain
# Course Set up 

---
class: left
# Course Set up - cont'd

The initial course set up is outlined [here](https://github.com/bcb420-2020/General_Course_Info/wiki#Initial_set_up) and will count for 5 marks of your grade.  It is due **Friday January 10,2020 by 5:00 PM**

## Step 1 

In order to get started you need to get added to the [bcb420-2020 organization](https://github.com/bcb420-2020).&lt;br&gt;
Enter an  issue:&lt;br&gt;&lt;li&gt;https://github.com/bcb420-2020/Student_Wiki/issues&lt;/li&gt; -  containing your github username that needs to be added to the organization&lt;br&gt;.

---
#Course Set up - cont'd

## Step 2

Everyone in the course will have a repository in the [bcb420-2020 organization](https://github.com/bcb420-2020) that belongs to them.  This repo is where you will submit all your assignments and maintain your own journal as a wiki associated with the repo.&lt;br&gt;  
Once your repo has been set up you need to:&lt;br&gt;
1. Start your course journal - there is a sample student repo with a sample journal posting that can be found [here](https://github.com/bcb420-2020/ExampleStudent/wiki).
1.  Start going through the preparatory material - start [here](https://bcb420-2020.github.io/General_course_prep/)
1.  Follow the remaining steps outlined in the [initial assignment](https://github.com/bcb420-2020/General_Course_Info/wiki#Initial_set_up)

---
# Course Set up - cont'd

## Step 3
* Look through the information on the main course page - [Main course website](https://github.com/bcb420-2020/General_Course_Info/wiki) 
* Make sure you add the links to your journal  to the [Main Student wiki](https://github.com/bcb420-2020/Student_Wiki/wiki) 

---
class: center
# Course discussions

&lt;font size=10&gt;Which platform do you prefer to have class discussions in?&lt;/font&gt; &lt;br&gt;&lt;br&gt;

&lt;font size=10&gt;Google groups or Quercus &lt;/font&gt; &lt;br&gt;&lt;br&gt;

&lt;font size=10&gt;fill in the survey here - https://forms.gle/QAfcA56EbgiFx4Hs9 &lt;/font&gt; &lt;br&gt;&lt;br&gt;

---
class: left
# Quiz

&lt;font size=12&gt; The quiz is worth 20 marks&lt;/font&gt;

The purpose of the quiz is to make sure we are all starting on the same page.  You are expected to already be familiar with R so this should just be a refresher.  &lt;br&gt;&lt;br&gt;
All the [preparatory material](https://github.com/bcb420-2020/General_Course_Info/wiki#Preparatory_Materials_and_Prerequisites) will be on the quiz.&lt;br&gt;&lt;br&gt;  
Going through all the tasks in https://bcb420-2020.github.io/R_basics/ is good preparation for the quiz. &lt;br&gt;&lt;br&gt;
We will have a mock quiz next week, **January 14, 2020 during class**.  We will then review and discuss the quiz after it is complete so everyone knows what to expect for the real quiz on **January 21, 2020 during class**

---
class: left
# Wiki

* In past iterations of this course the wiki used was hosted on [Prof. Steipe's personal wiki](http://steipe.biochemistry.utoronto.ca/abc/index.php/Computational_Systems_Biology_Main_Page). **Please feel free to look at the material that is there. There isa lot of great background and refresher information there.**
* We will be using github wikis for this course.  **They do not have the same rich feature set that is available in mediawiki** 
* Every student will maintain their own wiki in their own repository but you can also add links and content to the main [Student wiki](https://github.com/bcb420-2020/Student_Wiki/wiki) and are encouraged to do so. 
* There is a section in the preparatory material that discusses [Wiki](https://bcb420-2020.github.io/General_course_prep/wiki.html).  It is geared to MediaWiki.  You can set github wiki to recognize the MediaWiki format but it might not recognize all features (for example, templates.  I placed the copyright information as a footer of my wiki instead. )

---
class: left
# Journal

* **Main purpose** : to develop good habits. 
* Things to remember:
  * Often the person we are writing notes for is our future selves when we revisit a project, need to write up all the details of a given project for publication.
  * data transformations, parameters, code version are good details to include.
  * Errors that you encountered and how you fixed them! **They will come up again.  I guarentee it!**
* **Course Journal**:
  * Put your journal in the wiki of your course repo.
  * Make each entry its own page in the wiki.
  * Give the page a meaningful name
  * On the main page of your wiki link to all your added journal pages.  Feel free to organise them as you see fit but remember we will be looking through these notes so don't just put journal entry 1, journal entry 2.... 
* See [journal course prepratory material](https://bcb420-2020.github.io/General_course_prep/journal.html) for more details and template of a journal entry. 

---
class: left
# Assignment #1

* **Due January 28,2020 **
* Outline on the course wiki [here](https://github.com/bcb420-2020/General_Course_Info/wiki/Assignment%231)&lt;br&gt;&lt;br&gt;
* **Overview** : 
  * &lt;font size=12&gt;Choose a dataset,&lt;/font&gt; 
  * &lt;font size=12&gt;get to know it,&lt;/font&gt;  
  * &lt;font size=12&gt;claim it, and &lt;/font&gt; 
  * &lt;font size=12&gt;clean it&lt;/font&gt; &lt;br&gt;&lt;br&gt;
* We will be going over an example dataset in week 2 and week 3 of class as well. 

---

class: left
# Example Scenario (based on true events)
  * Imagine this situation - 
    * You are processing some expression dataset and you are 95% complete 
--

    * In a seminar you recently attended you saw a new way to visualize the same data that you are using that looked amazing!  
--

    * And who doesn't love a pretty visual representation.  
--

    * You want to add a new plot to your current pipeline.&lt;br&gt;
--

  * When you go to install **package_new** unfortunately it says that you need a newer version of R...&lt;br&gt;
--

  * But you really want to add this new plot! **Should you update R in order to install this new package?**&lt;br&gt;
--

.center[&lt;img src="https://media.giphy.com/media/RrVzUOXldFe8M/giphy.gif" width="25%" /&gt;]
  
---
class: left
# Example Scenario - cont'd

  * You install a newer version of R.&lt;br&gt;
--

.center[&lt;img src="https://media.giphy.com/media/pa37AAGzKXoek/giphy.gif" width="25%" /&gt;]
--

  * You run your current pipeline again.&lt;br&gt;
--

  * Sorry can't find package12 or package21 ....&lt;br&gt;
--

.center[&lt;img src="https://media.giphy.com/media/tJeGZumxDB01q/giphy.gif" width="25%" /&gt;]

---
class: left
# Example Scenario - cont'd

  * You notice that you never added the install statements to your script for those 2 packages that you added a few months back that helps with process x of your pipeline.  

--
  * You add the install statements now.  

--

  * **See it was good that I updated R.  Caught some bugs**

.center[&lt;img src="https://media.giphy.com/media/rIq6ASPIqo2k0/giphy.gif" width="25%" /&gt;]

---
class: left
# Example Scenario - cont'd

  * You re-run your pipeline with your added **package-new** and your new beautiful cutting edge plot and everything runs smoothly.

--

.center[&lt;img src="https://media.giphy.com/media/pa37AAGzKXoek/giphy.gif" width="25%" /&gt;]

--
  * You go through your original results but something doesn't look right!

--
  * **Those weren't your results!** 

--
.center[&lt;img src="https://media.giphy.com/media/11tTNkNy1SdXGg/giphy.gif" width="50%" /&gt;]

---
class: left 

# What happened

What could have happened here? **Ideas**?&lt;br&gt;

--

  * you installed the latest package-y but were previously using version 1.2.  
  * from version 1.2 to 1.3 they made a change to algorithm x that changed your results slightly.&lt;br&gt;
  
--

Truthfully lots could have happened. Especially when you are using multiple packages and multiple parts of your analysis.  

---
class: left 

# How could we have done this differently?

* Any ideas?

---
class: center, middle

&lt;img src="https://www.docker.com/sites/default/files/d8/styles/role_icon/public/2019-07/vertical-logo-monochromatic.png?itok=erja9lKc" /&gt;

---

class: left

#Docker
* What is Docker?&lt;br&gt;

--

  * [Docker](https://www.docker.com/) is a container platform similiar to the virtual machine but better. &lt;br&gt;

--

* Why is Docker **great** for Bioinformatics?&lt;br&gt;
  * allows you to create environments to run bioinformatis pipelines.&lt;br&gt;
  
--
  * create a consist environment to use for your pipelines. &lt;br&gt;

--
  * test modifications to the pipeline without disrupting your current set up.&lt;br&gt;
  

---
class: left, middle

.pull-left[
![](./images/docker_standalone.png)
]

--

.pull-right[
* Expensive
* Moving App to new system requires build a whole new server
* Potential wasted resources
* When you out grow the system is difficult to add resources.
]

.footnote[
*slide based [Introduction to docker - 2017](https://www.slideshare.net/Docker/introduction-to-docker-2017) slideshare (slide 16-26)
]
---

class: left, middle

.pull-left[
![](./images/docker_vm.png)
]

--

.pull-right[
**Pros**&lt;br&gt;
* Better resource usage
* easier to scale
* encapsulation of app/website&lt;br&gt;
&lt;br&gt;
**Cons**&lt;br&gt;
* Each VM still needs to be allocated CPUs, RAM, storage, operating system
* The more VMs you have the more resources you need.
* Portability not guarenteed but definitely easier than the old school individual server model
]

.footnote[
*slide based [Introduction to docker - 2017](https://www.slideshare.net/Docker/introduction-to-docker-2017) slideshare (slide 16-26)
]
---

class: left, middle

.pull-left[
![](https://docs.docker.com/images/VM%402x.png)
]

.pull-right[
![](https://docs.docker.com/images/Container%402x.png)
]

A VM encapsulates a server  whereas a Container encapsulates an application or service.

.footnote[
* from the docker [getting started guide](https://docs.docker.com/get-started/)
]

---

# Crash course on docker

Where to start:
  * Docker - [getting started guide](https://docker-curriculum.com/)
  * [Docker hub](https://hub.docker.com/) - repository of docker images
  
Instances we will be using:
  * Rocker - a docker instance configured to run R and Rstudio and various other flavours.  There are different base packages that contain different sets of R packages as default
    * [main rocker page](https://www.rocker-project.org/)
    * [rocker github repo](https://github.com/rocker-org/rocker)
    * [rocker on docker hub](https://www.rocker-project.org/)
  * Bioc - a docker instance configured to run R and contain different sets of commonly used Bioconductor packages.
    * [Main bioc docker page](https://www.bioconductor.org/help/docker/)
    * [bioconductor github repo](https://github.com/Bioconductor/bioc_docker) - there is some good starter notes here as well on how to set up your bioc container.
    * [bioconductor on docker hub](https://hub.docker.com/u/bioconductor/)

---

class: left

# Docker Basics

## Container
  * An instance of an image.
  * There can be multiple containers derived from the same image.
  * if you need to update it simply rebuild the image and relaunch a new container.

--

## Image
  * An image is basically the container blueprint.  It contains everything the container needs to run.

---
class: left

# Docker Basics - cont'd

## Dockerfile
  * A description of everything that is needed for the image
  * A set of commands to create the image
  * Contains commands:
    * FROM - which operatiing system to start from.  What is your base.
    * RUN - execute a given command
    * COPY - copy local files into the new container
    * CMD - to specify default parameters for the container
    * ENTRYPOINT - specify scripts to be run at image creation

---
class: left

# Example Dockerfile 


```r
*FROM bioconductor/release_core2:R3.6.0_Bioc3.9

# Install required packages
RUN install2.r \
    -d TRUE -e \
    -r "https://cran.rstudio.com" \
    -r "http://www.bioconductor.org/packages/release/bioc" \
    TCGAbiolinks \
    RTCGAToolbox 

#install additional packages
RUN install2.r \
   -d TRUE \
   -r "https://cran.rstudio.com" \
   -r "http://www.bioconductor.org/packages/release/bioc" \
   RColorBrewer@1.1-2 \
   GEOquery@2.52.0 \
   BiocGenerics@0.30.0
```

--
* Define the operating system or the base to build upon.

---
class: left

# Example Dockerfile - cont'd


```r
FROM bioconductor/release_core2:R3.6.0_Bioc3.9 

*# Install required packages
*RUN install2.r \
*   -d TRUE -e \
*   -r "https://cran.rstudio.com" \
*   -r "http://www.bioconductor.org/packages/release/bioc" \
*   TCGAbiolinks \
*   RTCGAToolbox

#install additional packages
RUN install2.r \
   -d TRUE \
   -r "https://cran.rstudio.com" \
   -r "http://www.bioconductor.org/packages/release/bioc" \
   RColorBrewer@1.1-2 \
   GEOquery@2.52.0 \
   BiocGenerics@0.30.0
```

--
* Install latest versions of a given pakage.

---

# Example Dockerfile - cont'd


```r
FROM bioconductor/release_core2:R3.6.0_Bioc3.9 

# Install required packages
RUN install2.r \
    -d TRUE -e \
    -r "https://cran.rstudio.com" \
    -r "http://www.bioconductor.org/packages/release/bioc" \
    TCGAbiolinks \
    RTCGAToolbox 

#install additional packages
*RUN install2.r \
*  -d TRUE \
*  -r "https://cran.rstudio.com" \
*  -r "http://www.bioconductor.org/packages/release/bioc" \
*  RColorBrewer@1.1-2 \
*  GEOquery@2.52.0 \
*  BiocGenerics@0.30.0
```

--
* Install specific version of a package.  

---
class: left

# Docker Basics - Cont'd

## Volumes
  * Files written to a directory on the container will not remain on the container once it has been stopped.  
  * But keeping files between runs and sharing information between containers is very useful. 
  * volumes offer a mechanisms to store results from a container.  Mapping a drive from the host machine to a drive on the container.  Files written to the mapped drive on the container will persist on the host machine.
  * Volumes are defined at runtime when the container is initially created.
  

--

```r
docker run -e PASSWORD=bcb420_pass 
    -p 8787:8787 
*   -v "$(pwd)"/bcb420_course_work:/home/rstudio/projects
    bcb420_image_base
```

---
class: left

## Port forwarding

* By default the container is self contained, with no communication with the host or any other containers.
* In order to open communication you can map a port in the container to the host.

--

```r
docker run -e PASSWORD=bcb420_pass 
*   -p 8787:8787
    -v "$(pwd)"/bcb420_course_work:/home/rstudio/projects 
    bcb420_image_base
```

--

* To access the container:
  * Open a browser and go to localhost:8787
  
---
class: left

# Conclusions
  * If we return to our starting example.  How would docker have helped the situation?
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
